From 5b507d99e05a152699ba44d95ea1ec26cb221fb9 Mon Sep 17 00:00:00 2001
From: Chilledheart <rwindz0@gmail.com>
Date: Tue, 17 Jan 2017 15:09:39 +0800
Subject: [PATCH 1/2] scan timed out cmd in nop out

---
 libiscsi-1.18.0/lib/nop.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/libiscsi-1.18.0/lib/nop.c b/libiscsi-1.18.0/lib/nop.c
index 6d9f450..065be42 100644
--- a/libiscsi-1.18.0/lib/nop.c
+++ b/libiscsi-1.18.0/lib/nop.c
@@ -33,12 +33,15 @@ iscsi_nop_out_async(struct iscsi_context *iscsi, iscsi_command_cb cb,
 	struct iscsi_pdu *pdu;

 	if (iscsi->old_iscsi || iscsi->pending_reconnect) {
+		if (iscsi->old_iscsi) iscsi_timeout_scan(iscsi->old_iscsi);
 		ISCSI_LOG(iscsi, (iscsi->nops_in_flight > 1) ? 1 : 6,
 		    "NOP Out Send NOT SEND while reconnecting (nops_in_flight: %d, iscsi->maxcmdsn %08x, iscsi->expcmdsn %08x)",
 		    iscsi->nops_in_flight, iscsi->maxcmdsn, iscsi->expcmdsn);
 		return 0;
 	}

+	iscsi_timeout_scan(iscsi);
+
 	if (iscsi->is_loggedin == 0) {
 		iscsi_set_error(iscsi, "trying to send nop-out while not "
 				"logged in");
--
2.9.3

